runtime/cudaq/dynamics/boson_operators.cpp

```diff
@@ -0,0 +1,215 @@
+  auto it = dimensions.find(this->target);
+  if (it == dimensions.end())
+    throw std::runtime_error("missing dimension for degree " + std::to_string(this->target));
+  return this->op_code_to_string();
+}
+
+void boson_operator::inplace_mult(const boson_operator &other) {
+  this->number_offsets.reserve(this->number_offsets.size() + other.number_offsets.size());
+
+  // first permute all number operators of RHS to the left; for x = # permutations,
+  // if we have "unpaired" creation operators, the number operator becomes (N + x),
+  // if we have "unpaired" annihilation operators, the number operator becomes (N - x).
+  auto it = this->number_offsets.cbegin(); // we will sort the offsets from biggest to smallest
+  for (auto offset : other.number_offsets) {
+    while (it != this->number_offsets.cend() && *it >= offset - this->additional_terms) ++it;
+    this->number_offsets.insert(it, offset - this->additional_terms);
+  }
+
+  // now we can combine the creation and annihilation operators;
+  if (this->additional_terms > 0) { // we have "unpaired" creation operators
+    // using ad*a = N and ad*N = (N - 1)*ad, each created number operator has an offset 
+    // of -(x - 1 - i), where x is the number of creation operators, and i is the number 
+    // of creation operators we already combined
+    it = this->number_offsets.cbegin();
+    for (auto i = std::min(this->additional_terms, -other.additional_terms); i > 0; --i) {
+      // we make sure to have offsets get smaller as we go to keep the sorting cheap
+      while (it != this->number_offsets.cend() && *it >= i - this->additional_terms) ++it;
+      this->number_offsets.insert(it, i - this->additional_terms);
+    }
+  } else if (this->additional_terms < 0) { // we have "unpaired" annihilation operators
+    // using a*ad = (N + 1) and a*N = (N + 1)*a, each created number operator has an offset 
+    // of (x - i), where x is the number of annihilation operators, and i is the number 
+    // of annihilation operators we already combined
+    it = this->number_offsets.cbegin();
+    for (auto i = 0; i > this->additional_terms && i > -other.additional_terms; --i) {
+      // we make sure to have offsets get smaller as we go to keep the sorting cheap
+      while (it != this->number_offsets.cend() && *it >= i - this->additional_terms) ++it;
+      this->number_offsets.insert(it, i - this->additional_terms);
+    }
+  }
+
+  // finally, we update the number of remaining unpaired operators
+  this->additional_terms += other.additional_terms;
+
+#if !defined(NDEBUG)
+  // we sort the number offsets, such that the equality comparison and the operator id
+  // perfectly reflects the mathematical evaluation of the operator
+  auto sorted_offsets = this->number_offsets;
+  std::sort(sorted_offsets.begin(), sorted_offsets.end(), std::greater<int>());
+  //assert(sorted_offsets == this->number_offsets);
```

Comment:
Q; Do we want to add this `assert` back for the debug build? Otherwise, I guess we can remove this whole `!defined(NDEBUG)` block.

Reply:
good catch. I find the debug assertions helpful to immediately detect bugs in the tests when something about the logic is clearly not doing what was intended.